src/droplet.F

```diff
@@ -1234,268 +1123,563 @@ subroutine rk2_integration(dt,np,pdata,pdata_locind,iflag,jflag,kflag,xf,xh,uh,r
         vrhs2 = 1.0/taup*(vval - pdata(np,prvpy)) + part_grav2
         vrhs3 = 1.0/taup*(wval - pdata(np,prvpz)) + part_grav3
 
-        ! einf = qinf*rhoval/mw*ru*tval
+      !Position and velocity are straightforward:
 
-        qstar = qinf*exp(                                                   &
-                     mw*lhv/ru*(1.0/tval - 1.0/pdata(np,prtp))              &
-                    +2.0*ms*surften/(ru*rhow*pdata(np,prrp)*pdata(np,prtp)) &
-                    -ion*os*pdata(np,prms)*mw/ms/(volp*rhop-pdata(np,prms)) &
-                        )
+      x3d = pdata(np,prx) + dt*pdata(np,prvpx)
+      y3d = pdata(np,pry) + dt*pdata(np,prvpy)
+      z3d = pdata(np,prz) + dt*pdata(np,prvpz)
 
-        rprhs = Shp/9.0/sc_num*rhop/rhow*pdata(np,prrp)/taup*(qval-qstar)
+      pdata(np,prvpx) = (pdata(np,prvpx) + dt*uval/taup + dt*part_grav1)/(1 + dt/taup)
+      pdata(np,prvpy) = (pdata(np,prvpy) + dt*vval/taup + dt*part_grav2)/(1 + dt/taup)
+      pdata(np,prvpz) = (pdata(np,prvpz) + dt*wval/taup + dt*part_grav3)/(1 + dt/taup)
 
-        tprhs = Nup/3.0/pr_num*(cp/cpl)*(rhop/rhow)/taup*(tval-pdata(np,prtp))+ &
-                3.0*lhv/cpl/pdata(np,prrp)*rprhs
+      !Nonlinear solver for the radius and temperature:
 
+      !Nondimensionalize parameters before the solver using 1-micron water droplet as ref
+      taup_scale = rhow*(1.0e-6)**2/(18*rhoval*viscosity)
 
-        IF(nx.eq.1)THEN
-          x3d=0.0
-          vtmp1 = 0.0
-        ELSE
-          k1xp1 = dt*xrhs1
-          x3d=pdata(np,prx)+k1xp1
+      dt_nondim = dt/taup_scale
+      
 
-          k1vp1 = dt*vrhs1
-          vtmp1 = pdata(np,prvpx) + k1vp1
+      !Gives initial guess into nonlinear solver
+      !mflag = 0, has equilibrium radius; mflag = 1, no equilibrium (uses itself as initial guess)
+      call rad_solver2(guess,mflag,prsval,rhoval,tval,qval,pdata(np,prtp),pdata(np,prms),pdata(np,prrp))
+
+      if (mflag.eq.0) then
+         rt_start(1) = guess/pdata(np,prrp)
+         rt_start(2) = tval/pdata(np,prtp)
+      else
+         rt_start(1) = 1.0
+         rt_start(2) = 1.0
+      end if
+
+
+      call gauss_newton_2d(diffnorm,dt_nondim,taup_scale,rt_start,rt_zeros,flag,prsval,rhoval,tval,qval,pdata(np,prtp),pdata(np,prms),pdata(np,prrp))
+
+      if (flag==1) then
+         num100 = num100+1
+
+         call LV_solver(diffnorm,dt_nondim,taup_scale,rt_start,rt_zeros,flag,prsval,rhoval,tval,qval,pdata(np,prtp),pdata(np,prms),pdata(np,prrp))
+
+      end if
+
+      if (flag==1) num1000 = num1000+1
+
+      !In rare cases the nonlinear solvers fail to converge
+      if (isnan(rt_zeros(1)) .OR.  isnan(rt_zeros(2)) .OR. rt_zeros(2)<0) then
```

Comment:
According to this thread (https://forums.developer.nvidia.com/t/error-isnan-has-not-been-explicitly-declared/132718/6), `isnan` is not a standard extension and may not be supported by some compilers like `PGI`. I would suggest using a better way to detect the NaN value here.

Reply:
Has the isnan() issue been resolved yet?

Reply:
Not yet. I plan to work on this after we have a stable verification framework and namelist.

Reply:
Updated by @drichte2